### PR TITLE
Fix ssl_do_config and SSL_load_client_CA_file_ex changing library context

### LINKED ISSUE #14593: SSL_CTX_new is unexpectedly changing the current library context
`SSL_CTX_new()` and `SSL_CTX_new_ex()` does change the current library context to the initial default library context under some circumstances:

`SSL_CTX_new_ex()` calls `ssl_ctx_system_config()` which in turn calls `ssl_do_config()`. In `ssl_do_config()`, somewhere in the middle of the function, it uses `OSSL_LIB_CTX_set0_default()` to set the current library context to the one from the SSL context, and remembers the previous library context in `prev_libctx`. Further down (at the `err` label) it restores the previous library context to what it was before (i.e. `prev_libctx`).

So far so good, but if the function fails early, i.e. before initially setting the library context via `OSSL_LIB_CTX_set0_default()`, it did not change the current library context, and thus did not remember the previous context,  but at the `err` label it unconditionally restores the previous library context to `prev_libctx`, which is then NULL.  

However, calling `OSSL_LIB_CTX_set0_default()` with NULL changes the current library context to the initial default library context. So if a different library context was set up before and set as current one, it is now no longer the current one. It should not be assumed that calling `OSSL_LIB_CTX_set0_default` with NULL would be a no-op.

In my case it is failing at `conf_ssl_name_find()` to find the `"system_default"`, for whatever reason. It then goes to the `err` label and thus changes the current library context to NULL. Since I loaded my own provider only into my own library context, it won't find my provider from now on.

This problem was introduced with commit 6725682d77510bf6d499957897d7be124d603f40 as far as I can tell.

There are some other similar places with similar invalid restoring of the default library context:
- `SSL_load_client_CA_file_ex()` (ssl/ssl_cert.c) unconditionally restores the library context at the done label.  If the function fails early, it also calls `OSSL_LIB_CTX_set0_default`  with NULL (also since commit 6725682d77510bf6d499957897d7be124d603f40).
- `ASYNC_start_job()` (crypto/async/asyc.c) sets and restores the library context, but if `async_fibre_swapcontext()` fails, it did change the library context, but does not restore the previous one at the err label (since commit 6c689e58f70726cfa5533bc10f9669abce199fb8).

### PR DESCRIPTION
**Pull request by ifranzki**
Do not change the current default library context when NULL is passed as the library context to use. Also do not restore the previous library context when it was not changed before.

Fixes #14593

##### Checklist

- [ ] documentation is added or updated
- [ ] tests are added or updated

### CODE REVIEW

File: ssl/ssl_cert.c [hunk 1]
```diff
@@ -664,7 +665,8 @@ STACK_OF(X509_NAME) *SSL_load_client_CA_file_ex(const char *file,
     ret = NULL;
  done:
     /* restore the old libctx */
-    OSSL_LIB_CTX_set0_default(prev_libctx);
+    if (prev_libctx != NULL)
```
Comment (mattcaswell) [inline ssl/ssl_cert.c:668]: Shouldn't this be `if (libctx != NULL)` here?

File: ssl/ssl_mcnf.c [hunk 2]
```diff
@@ -81,7 +82,8 @@ static int ssl_do_config(SSL *s, SSL_CTX *ctx, const char *name, int system)
     }
     rv = SSL_CONF_CTX_finish(cctx);
  err:
-    OSSL_LIB_CTX_set0_default(prev_libctx);
+    if (prev_libctx != NULL)
```
Comment (mattcaswell) [inline ssl/ssl_mcnf.c:85]: Same here

**Review by mattcaswell — CHANGES_REQUESTED** [review]

File: ssl/ssl_cert.c (hunk 1 above)
Comment (t8m) [reply ssl/ssl_cert.c:668]: No. `prev_libctx == NULL && libctx != NULL` can happen only if OSSL_LIB_CTX_set0_default() fails which means the prev_libctx is not meaningful anyway. And there are code paths which can get here where the OSSL_LIB_CTX_set0_default() was not called at all.

File: ssl/ssl_mcnf.c (hunk 2 above)
Comment (t8m) [reply ssl/ssl_mcnf.c:85]: Same answer as above

File: ssl/ssl_cert.c (hunk 1 above)
Comment (mattcaswell) [reply ssl/ssl_cert.c:668]: Logically I find this strange. In my mind the condition for checking whether to set a new default lib ctx should be the same for deciding whether to reset it on exit. It is odd to use different conditions IMO.

Either way, should we be testing for prev_libctx being NULL after the first `OSSL_LIB_CTX_set0_default()` and erroring out in that case?

Comment (t8m) [reply ssl/ssl_cert.c:668]: > Either way, should we be testing for prev_libctx being NULL after the first `OSSL_LIB_CTX_set0_default()` and erroring out in that case?

Yes, that would make sense. But this condition should be still kept.

**Comment by t8m** [comment]
Now I do not think this is covered by CLA: trivial. Could you please submit CLA? https://www.openssl.org/policies/cla.html

**Comment by ifranzki** [comment]
> Now I do not think this is covered by CLA: trivial. Could you please submit CLA?

I feared that this would be required.... This will take some time due to company internal processes.

I will ask @p-steuer (Patrick Steuer) to submit a PR for this fix, he has a CLA.

**Comment by ifranzki** [comment]
Or just submit a PR on your own, if you don't want to wait....

**Comment by t8m** [comment]
> Or just submit a PR on your own, if you don't want to wait....

There is no hurry IMO.

**Comment by ifranzki** [comment]
My Individual CLA as well as the company CLA should now be on file. We have received the following from the OpenSSL legal team: 

> "I've add Ingo to our database, he should be good to contribute now.  Only open pull requests will need to be closed and reopened for the various hold tages to disappear."

How do you want me to continue o this PR? Should I close this one and open a new one, or do you want me to rebase the commits and remove the CLA: trivial from it?

**Comment by t8m** [comment]
Please rebase and remove the CLA: trivial from the commits.

**Review by t8m — APPROVED** [review]

**Review by mattcaswell — APPROVED** [review]

**Comment by mattcaswell** [comment]
Hmmm.....actually....I'm just wondering whether this problem could be better solved in `OSSL_LIB_CTX_set0_default`, e.g. something like:

````
diff --git a/crypto/context.c b/crypto/context.c
index 852f9dd7ce..e627141006 100644
--- a/crypto/context.c
+++ b/crypto/context.c
@@ -188,9 +188,11 @@ OSSL_LIB_CTX *OSSL_LIB_CTX_set0_default(OSSL_LIB_CTX *libctx)
 #ifndef FIPS_MODULE
     OSSL_LIB_CTX *current_defctx;
 
-    if ((current_defctx = get_default_context()) != NULL
-        && set_default_context(libctx))
+    if ((current_defctx = get_default_context()) != NULL) {
+        if (libctx != NULL)
+            set_default_context(libctx);
         return current_defctx;
+    }
 #endif
 
     return NULL;
````

Otherwise are we going to have to potentially do this dance with a NULL libctx everywhere that we call OSSL_LIB_CTX_set0_default()?

**Comment by mattcaswell** [comment]
Sorry...didn't mean to click close...

**Comment by t8m** [comment]
> Hmmm.....actually....I'm just wondering whether this problem could be better solved in `OSSL_LIB_CTX_set0_default`

This makes some sense, however how do you then forcibly set the default default context with OSSL_LIB_CTX_set0_default()? We would need a different API call for that.

**Comment by mattcaswell** [comment]
> This makes some sense, however how do you then forcibly set the default default context with OSSL_LIB_CTX_set0_default()? We would need a different API call for that.

Why is that necessary? You can of course always do that if you have previously called `OSSL_LIB_CTX_set0_default()` when the default default context was already set and saved away the return value - which is the intended way that `OSSL_LIB_CTX_set0_default()` is supposed to be used.

**Comment by ifranzki** [comment]
If you go that way, then you should clearly document that `OSSL_LIB_CTX_set0_default()` when called with NULL does nothing, but just returns the current context. 
You should also check all current callers of `OSSL_LIB_CTX_set0_default()` if it then still behaves as desired.

**Comment by t8m** [comment]
> Why is that necessary?

I do not have a concrete use-case in mind. But I suspect someone will need this.

**Comment by t8m** [comment]
I would simply merge this and contemplate about changing OSSL_LIB_CTX_set0_default() in a separate issue. What do you think, Matt?

**Comment by mattcaswell** [comment]
I raised #14890 as an alternative to this PR.

**Comment by mattcaswell** [comment]
#14890 got merged, so I think this PR is no longer necessary. Thanks for your contribution anyway.

**Pull request closed without merging**